survey/src/main/java/com.quickbirdstudios.surveykit/backend/views/question_parts/LocationPickerPart.kt

```diff
@@ -0,0 +1,268 @@
+                )
+            } else {
+                mapReadySelection = selected
+            }
+        }
+
+    //endregion
+
+    //region Overrides
+
+    override fun style(surveyTheme: SurveyTheme) {}
+
+    //endregion
+
+    //region Subtypes
+
+    data class Selection(val latitude: Double?, val longitude: Double?)
+
+    //endregion
+
+    init {
+        this.orientation = VERTICAL
+        this.gravity = Gravity.CENTER
+
+        mapView = MapView(
+            context,
+            GoogleMapOptions().scrollGesturesEnabled(true)
+                .tiltGesturesEnabled(true).rotateGesturesEnabled(false)
+        )
+
+        search = AutoCompleteTextView(context)
+        search.setHint(android.R.string.search_go)
+        search.setSingleLine(true)
+        search.setCompoundDrawables(
+            resources.getDrawable(R.drawable.ic_baseline_search_24),
+            null,
+            null,
+            null
+        )
+        search.compoundDrawablePadding = 32
+
+        mapView.getMapAsync { map ->
+            this.map = map
+            if (::mapReadySelection.isInitialized) {
+                selected = mapReadySelection
+            }
+
+            if (ActivityCompat.checkSelfPermission(
+                    mapView.context,
+                    Manifest.permission.ACCESS_FINE_LOCATION
+                ) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
+                        mapView.context,
+                        Manifest.permission.ACCESS_COARSE_LOCATION
+                    ) != PackageManager.PERMISSION_GRANTED
+            ) {
+                // TODO ask location permission
+                return@getMapAsync
+            }
+            this.map.isMyLocationEnabled = true
+            this.map.uiSettings.isMyLocationButtonEnabled = true
+
+            this.map.setOnMyLocationChangeListener { location ->
+                if (selected.latitude == 0.0 && selected.longitude == 0.0) {
+                    this.map.moveCamera(
+                        CameraUpdateFactory.newLatLngZoom(
+                            LatLng(
+                                location.latitude,
+                                location.longitude
+                            ),
+                            15f
+                        )
+                    )
+                }
+            }
+
+            this.map.setOnCameraMoveStartedListener { reason ->
+                if (reason == GoogleMap.OnCameraMoveStartedListener.REASON_GESTURE) {
+                    locationMarker.animate().translationY(-10f)
+                        .setInterpolator(BounceInterpolator()).setDuration(300)
+                        .start()
+                }
+            }
+
+            this.map.setOnCameraIdleListener {
+                locationMarker.animate().translationY(10f)
+                    .setInterpolator(BounceInterpolator()).setDuration(300)
+                    .start()
+            }
+        }
```

Comment:
To avoid repetition of `this.map`  you can enter everything in a `with` block:

```kotlin
with(this.map){
    isMyLocationEnabled = true
    uiSettings.isMyLocationButtonEnabled = true

setOnMyLocationChangeListener { location ->
                if (selected.latitude == 0.0 && selected.longitude == 0.0) {
                    this.map.moveCamera(
                        CameraUpdateFactory.newLatLngZoom(
                            LatLng(
                                location.latitude,
                                location.longitude
                            ),
                            15f
                        )
                    )
                }
            }
...
}
```